nemo/nemo/collections/nlp/modules/common/megatron/transformer.py

```diff
@@ -1116,6 +1146,19 @@ def forward(
         if get_key_value:
             present = (key_layer, value_layer)
 
+        # expand the key_layer and value_layer [sk, b, ng, hn] -> [sk, b, np, hn]
+        # This is a noop for normal attention where ng == np. When using grouped-query attention this
+        # creates a view that has the keys and values virtually repeated along their dimension to
+        # match the number of queries.
+        key_layer = key_layer.repeat_interleave(
```

Comment:
How is this repeat_intervleaving done? Does it use an explicit torch.view() operation?

Reply:
@amithrm 
Instead of reshaping a tensor like `torch.view`, it repeats elements of a tensor.

GQA does not have a one-to-one correspondence between query heads and key/value heads like MHA.
Instead, multiple query heads share a single key/value head.
By virtually repeating shared key/value heads until the number of heads becomes `num_attention_heads`, `core_attention` can treat MHA and GQA equivalently.

The middle and right illustrations in Figure 2 will be transformed to have the same shape as the left one through this operation.

<img width="805" alt="254449231-2808a8a9-2c5c-4c72-b7cf-bbdb4468832b" src="https://github.com/aws-neuron/neuronx-nemo-megatron/assets/21357689/008ce04c-9292-42e0-8e1c-9cf322664fb0">

Reply:
Wouldn't this cause calculations to be duplicated/redundant across GQA groups?

Reply:
This operation does not increase the time complexity.
GQA reduces the complexity by reducing the dimension of the projection layer that transforms hidden states into key/value heads.
The complexity of the dot product in the core attention is equivalent for both GQA and MHA.